addons/stock/models/stock_move.py

```diff
@@ -1144,6 +1145,9 @@ def _action_confirm(self, merge=True, merge_into=False):
 
         move_to_confirm.write({'state': 'confirmed'})
         (move_waiting | move_create_proc).write({'state': 'waiting'})
+        # procure_method sometimes changes with certain workflows so just in case, apply to all moves
+        self.filtered(lambda m: m.picking_type_id.reservation_method == 'at_confirm' and not m.reservation_date)\
```

Comment:
IMO, The `not m.reservation_date` should be remove because, I suppose that when you create a draft transfers with `picking_type_id.reservation_method == 'by_date'` and add a new draft move, save it.
The move will have a reservation_date. Then change the picking_type_id by one with `reservation_method == 'at_confirm'` and then it will be wrong at the moment you confirm your move (because the reservation_date will be set).

Reply:
Fair enough. Changed it. (also fixed my mistake of not only doing the moves having their states changed)